test/param_build_test.c

```diff
@@ -0,0 +1,251 @@
+        /* Check unsigned int32 */
+        || !TEST_ptr(p = OSSL_PARAM_locate(params, "i32"))
+        || !TEST_true(OSSL_PARAM_get_uint32(p, &i32))
+        || !TEST_str_eq(p->key, "i32")
+        || !TEST_uint_eq(p->data_type, OSSL_PARAM_UNSIGNED_INTEGER)
+        || !TEST_size_t_eq(p->data_size, sizeof(int32_t))
+        || !TEST_uint_eq((unsigned int)i32, 1532)
+        /* Check unsigned int64 */
+        || !TEST_ptr(p = OSSL_PARAM_locate(params, "i64"))
+        || !TEST_str_eq(p->key, "i64")
+        || !TEST_uint_eq(p->data_type, OSSL_PARAM_UNSIGNED_INTEGER)
+        || !TEST_size_t_eq(p->data_size, sizeof(int64_t))
+        || !TEST_true(OSSL_PARAM_get_uint64(p, &i64))
+        || !TEST_ulong_eq((unsigned long)i64, 9999999)
+        /* Check unsigned long int */
+        || !TEST_ptr(p = OSSL_PARAM_locate(params, "l"))
+        || !TEST_str_eq(p->key, "l")
+        || !TEST_uint_eq(p->data_type, OSSL_PARAM_UNSIGNED_INTEGER)
+        || !TEST_size_t_eq(p->data_size, sizeof(unsigned long int))
+        || !TEST_true(OSSL_PARAM_get_ulong(p, &l))
+        || !TEST_ulong_eq(l, 42)
+        /* Check size_t */
+        || !TEST_ptr(p = OSSL_PARAM_locate(params, "st"))
+        || !TEST_str_eq(p->key, "st")
+        || !TEST_uint_eq(p->data_type, OSSL_PARAM_UNSIGNED_INTEGER)
+        || !TEST_size_t_eq(p->data_size, sizeof(size_t))
+        || !TEST_true(OSSL_PARAM_get_size_t(p, &st))
+        || !TEST_size_t_eq(st, 65537)
+        /* Check octet string */
+        || !TEST_ptr(p = OSSL_PARAM_locate(params, "oct_s"))
+        || !TEST_str_eq(p->key, "oct_s")
+        || !TEST_uint_eq(p->data_type, OSSL_PARAM_OCTET_STRING)
+        || !TEST_mem_eq(p->data, p->data_size, data1, sizeof(data1))
+        /* Check octet pointer */
+        || !TEST_ptr(p = OSSL_PARAM_locate(params, "oct_p"))
+        || !TEST_str_eq(p->key, "oct_p")
+        || !TEST_uint_eq(p->data_type, OSSL_PARAM_OCTET_PTR)
+        || !TEST_mem_eq(*(void **)p->data, p->data_size, data2, sizeof(data2))
+        /* Check BN */
+        || !TEST_ptr(p = OSSL_PARAM_locate(params, "bignumber"))
+        || !TEST_str_eq(p->key, "bignumber")
+        || !TEST_uint_eq(p->data_type, OSSL_PARAM_UNSIGNED_INTEGER)
+        || !TEST_true(OSSL_PARAM_get_BN(p, &bn_res))
+        || !TEST_int_eq(BN_cmp(bn_res, bn), 0))
+        goto err;
+    res = 1;
+err:
+    OPENSSL_secure_free(params);
+    BN_free(bn);
+    BN_free(bn_res);
+    return res;
+}
+
+static int template_static_params_test(int n)
+{
+    unsigned char data[1000];
+    OSSL_PARAM_BLD bld;
+    OSSL_PARAM params[20], *p;
+    unsigned int i;
+    char *utf = NULL;
+    int res = 0;
+
+    ossl_param_bld_init(&bld);
+    if (n == 1)
+        ossl_param_bld_set_private(&bld);
+    if (!TEST_true(ossl_param_bld_push_uint(&bld, "i", 6))
+        || !TEST_true(ossl_param_bld_push_utf8_string(&bld, "utf8_s", "bar",
+                                                      0))
+        || !TEST_ptr(ossl_param_bld_to_param_ex(&bld, params,
```

Comment:
I was imagining that this API would be a bit simpler.. i.e:
unsigned char data[1000];
OSSL_PARAM *params;
...
ossl_param_bld_to_param_ex(&bld, params, data, sizeof(data));
...
(Where params just uses the data block)...

Reply:
So allocate the OSSL_PARAM array but not the storage?  The [original suggestion](https://github.com/openssl/openssl/pull/9305#discussion_r300834215) was for both to be allocated by the caller.

It would be possible to support both allocation of the OSSL_PARAM array and not.  It would be tricky to support allocation of the data without the param array (freeing something you don't have a reference to).

My thinking is that if the caller wants to allocate the data space, they might as well also allocate the param space.

Reply:
No I mean params points to the data block..

Reply:
Ah, everything is in the one block.  That could be done.  It wasn't what @levitte asked for but it would be possible -- data alignment might be tricky.

Reply:
I'm wondering if that puts too much of a burden on the caller.  They will not only have to calculate how much space the data is likely to take up, but will also have to add the space the params will take up.  It also means that `ossl_param_bld_to_param_ex` will have to make *assumptions* about where exactly `params` is pointing.

Please do try and see if you can get that done cleanly...  I have my doubts, though.

Reply:
all the time spent to avoid memory fragmentation in a small piece of a running application. :)

Reply:
It is more a question of ease of freeing -- one block means one free.

Reply:
still not an issue, you have a type-specific free function, just like, oh, ```RSA_free```, ```X509_free``` and the like.  The developers don't need to know any internals. Again, the only real impact of this is memory fragmentation, which seems the wrong thing to optimize for at this point. Instead of obvious code, you end up with tricky code.  Wrong trade-off.

Reply:
It can be done by adding the assumption that the passed in pointer is properly aligned for all types.  The `unsigned char data[1000]` example wouldn't meet this criteria.

Without this assumption, pointer alignment will need to be forced but is possible (union to portably convert to uintptr_t, some integer operations and a conversion back).

I'm not convinced that it is worth the effort.

Reply:
if OSSL_PARAM was padded then it would be easy, otherwise dont bother.